src/python/turicreate/toolkits/sound_classifier/sound_classifier.py

```diff
@@ -21,6 +21,82 @@
 
 USE_TF = _tk_utils._read_env_var_cpp('TURI_SC_USE_TF_PATH')
 
+class _DataIterator(object):
+    '''
+    Defines a common interface around MXNet NDArrayIter and TensorFlow
+    DataSet.from_tensor_slices
+    '''
+
+    '''
+    Creates a new instance wrapping numpy data and labels.
+
+    If `label` is provided, `label.shape[0]` must match `data.shape[0]`.
+    '''
+    def __init__(self, data, label=None, batch_size=1, shuffle=False):
+        raise NotImplementedError
+
+    '''
+    Returns an iterator that yields a sequence of tuples, comprising a batch of
+    `data` values and a batch of `label` values (if provided).
+    '''
+    def __iter__(self):
+        raise NotImplementedError
+
+    '''
+    Ensures that the next iteration through the dataset starts from the beginning.
+    '''
+    def reset(self):
+        raise NotImplementedError
+
+class _MXNetDataIterator(_DataIterator):
+    def __init__(self, data, label=None, batch_size=1, shuffle=False):
+        import mxnet as mx
+        self.impl = mx.io.NDArrayIter(data, label=label, batch_size=batch_size,
+                                      shuffle=shuffle)
+
+    def __iter__(self):
+        return self
+
+    def reset(self):
+        self.impl.reset()
+
+    def __next__(self):
+        batch = self.impl.__next__()
+        data = batch.data[0]
+        label = batch.label[0] if batch.label else None
+        if batch.pad != 0:
+            data = data[:-batch.pad]
+            label = label[:-batch.pad] if label is not None else None
+        data = data.asnumpy()
+        label = label.asnumpy() if label is not None else None
+        return (data, label) if label is not None else (data,)
+
+    def next(self):
+        return self.__next__()
+
+class _TFDataIterator(_DataIterator):
+    def __init__(self, data, label=None, batch_size=1, shuffle=False):
+        import tensorflow as tf
+        tensor_slices = (data, label) if label is not None else (data,)
+        self.impl = tf.data.Dataset.from_tensor_slices(tensor_slices)
+        self.impl = self.impl.batch(batch_size)
+        if shuffle:
+            self.impl = self.impl.shuffle(data.shape[0])
+
+    def __iter__(self):
+        return self.impl.__iter__()
+
+    def reset(self):
+        pass
```

Comment:
Do you want to raise an exception here? If not, it would be great to comment on why a `pass` here is ok.

Reply:
Good point. It turns out that TF (at least in 2.0) returns a fresh iterator every time for `__iter__()`, unlike MXNet, which uses the main instance itself as an iterator